### PR TITLE
stat: implement Pareto

### PR DESCRIPTION
**Pull request by kczimm**
An implementation of the Pareto (Type I) distribution without the Rand random sampling and the Quantile methods.

### CODE REVIEW

**Comment by kczimm** [comment]
Added the Rand function and associated tests. Still do not have the Quantile method implemented though.

File: stat/distuv/pareto.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	"math"
+
+	"golang.org/x/exp/rand"
+)
+
+// Pareto implements the Pareto (Type I) distribution, a one parameter distribution
+// with support above the scale parameter.
+//
+// The density function is given by
+// \frac{\alpha x_m^{\alpha}}{x^{\alpha+1}} for x >= x_m
```
Comment (kortschak) [inline stat/distuv/pareto.go]: We use something that is almost but not quite TeX for these formulae. In this case I'd not use `\frac` but rather a `/` and parens. Also, indent the formula by one space and finish with a dot.

File: stat/distuv/pareto.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+//
+// For more information, see https://en.wikipedia.org/wiki/Pareto_distribution
```
Comment (kortschak) [inline stat/distuv/pareto.go]: Final full stop.

File: stat/distuv/pareto.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+type Pareto struct {
+	// Xm is the scale parameter. Must be greater than 0.
+	Xm float64
+
+	// Alpha is the shape parameter. Must be greater than 0.
```
Comment (kortschak) [inline stat/distuv/pareto.go]: "Alpha must be greater than 0."

File: stat/distuv/pareto.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+// Pareto implements the Pareto (Type I) distribution, a one parameter distribution
+// with support above the scale parameter.
+//
+// The density function is given by
+// \frac{\alpha x_m^{\alpha}}{x^{\alpha+1}} for x >= x_m
+//
+// For more information, see https://en.wikipedia.org/wiki/Pareto_distribution
+type Pareto struct {
+	// Xm is the scale parameter. Must be greater than 0.
```
Comment (kortschak) [inline stat/distuv/pareto.go]: "Xm must be greater than 0."

File: stat/distuv/pareto.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	}
+	return 1 - math.Pow(p.Xm/x, p.Alpha)
+}
+
+// ExKurtosis returns the excess kurtosis of the distribution.
+func (p Pareto) ExKurtosis() float64 {
+	if p.Alpha <= 4 {
+		return 0
+	}
+	return 6 * (math.Pow(p.Alpha, 3) + math.Pow(p.Alpha, 2) - 6*p.Alpha - 2) / p.Alpha / (p.Alpha - 3) / (p.Alpha - 4)
```
Comment (kortschak) [inline stat/distuv/pareto.go]: s/math.Pow(p.Alpha, 3)/p.Alpha*p.Alpha*p.Alpha/
s/math.Pow(p.Alpha, 2)/p.Alpha*p.Alpha/
s|p.Alpha / (p.Alpha - 3) / (p.Alpha - 4)|(p.Alpha * (p.Alpha - 3) * (p.Alpha - 4))|

File: stat/distuv/pareto.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+func (p Pareto) LogProb(x float64) float64 {
+	if x < p.Xm {
+		return math.Inf(-1)
+	}
+	return math.Log(p.Alpha) + p.Alpha*math.Log(p.Xm) - (p.Alpha+1)*math.Log(x)
+}
+
+// Mean returns the mean of the probability distribution.
+func (p Pareto) Mean() float64 {
+	if p.Alpha > 1 {
```
Comment (kortschak) [inline stat/distuv/pareto.go]:
```
if p.Alpha <= 1 {
    return math.Inf(1)
}
return p.Alpha * p.Xm / (p.Alpha - 1)
```

File: stat/distuv/pareto.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+// NumParameters returns the number of parameters in the distribution.
+func (p Pareto) NumParameters() int {
+	return 2
+}
+
+// Prob computes the value of the probability density function at x.
+func (p Pareto) Prob(x float64) float64 {
+	if x < p.Xm {
```
Comment (kortschak) [inline stat/distuv/pareto.go]: Usually we `return math.Exp(p.LogProb(x))`.

File: stat/distuv/pareto.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+	}
+	return math.Pow(p.Xm/x, p.Alpha)
+}
+
+// Variance returns the variance of the probability distribution.
+func (p Pareto) Variance() float64 {
+	if p.Alpha <= 2 {
+		return math.Inf(1)
+	}
+	return p.Xm * p.Xm * p.Alpha / math.Pow(p.Alpha-1, 2) / (p.Alpha - 2)
```
Comment (kortschak) [inline stat/distuv/pareto.go]:
```
am1 := p.Alpha-1
return p.Xm * p.Xm * p.Alpha / (am1 * am1 * (p.Alpha - 2))
```

File: stat/distuv/pareto.go (hunk 7 above)
Comment (kczimm) [reply stat/distuv/pareto.go]: done

File: stat/distuv/pareto.go (hunk 2 above)
Comment (kczimm) [reply stat/distuv/pareto.go]: I don't know the proper lingo yet. What does final full stop mean? I put a trailing period here, but I'm not sure if that's what you meant.

Comment (kortschak) [reply stat/distuv/pareto.go]: I don't speak ASE, "full stop" == "period".

File: stat/distuv/pareto.go (hunk 7 above)
Comment (kortschak) [reply stat/distuv/pareto.go]: Omit the condition, you already check for `x < p.Xm` in `LogProb` and return `-Inf` in that case; `math.Exp(math.Inf(-1))` returns 0.

Comment (kczimm) [reply stat/distuv/pareto.go]: Ah, so obvious. Thank you. Done.

File: stat/distuv/pareto.go [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	"math"
+
+	"golang.org/x/exp/rand"
+)
+
+// Pareto implements the Pareto (Type I) distribution, a one parameter distribution
+// with support above the scale parameter.
+//
+// The density function is given by
+//  (\alpha x_m^{\alpha})/(x^{\alpha+1}) for x >= x_m.
```
Comment (btracey) [inline stat/distuv/pareto.go]: Could you swap `\alpha` with α

File: stat/distuv/pareto.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+	Src *rand.Rand
+}
+
+// CDF computes the value of the cumulative density function at x.
+func (p Pareto) CDF(x float64) float64 {
+	if x < p.Xm {
+		return 0
+	}
+	return 1 - math.Pow(p.Xm/x, p.Alpha)
```
Comment (btracey) [inline stat/distuv/pareto.go]: This should return `1 - p.Survival(x)`

File: AUTHORS [hunk 11]
```diff
@@ -38,6 +38,7 @@ Joseph Watson <jtwatson@linux-consulting.us>
 Josh Wilson <josh.craig.wilson@gmail.com>
 Julien Roland <juroland@gmail.com>
 Kent English <kent.english@gmail.com>
+Kevin C. Zimmerman <kevinczimmerman@gmail.com>
```
Comment (btracey) [inline AUTHORS]: We may not need these if the other one gets in first.

File: stat/distuv/pareto.go [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+// Mean returns the mean of the probability distribution.
+func (p Pareto) Mean() float64 {
+	if p.Alpha <= 1 {
+		return math.Inf(1)
+	}
+	return p.Alpha * p.Xm / (p.Alpha - 1)
+}
+
```
Comment (btracey) [inline stat/distuv/pareto.go]: Median? It's easy here.

File: stat/distuv/pareto.go [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+// Prob computes the value of the probability density function at x.
+func (p Pareto) Prob(x float64) float64 {
+	return math.Exp(p.LogProb(x))
+}
+
+// Rand returns a random sample drawn from the distribution.
+func (p Pareto) Rand() float64 {
+	return p.Xm / math.Pow(1-Uniform{Min: 0, Max: 1}.Rand(), 1/p.Alpha)
```
Comment (btracey) [inline stat/distuv/pareto.go]: 1) This doesn't need to construct a distribution, that's built into the *rand.Rand
2) It does need to use the rand.Rand to generate the variable if it's non-nil
3) This should work in log space instead 
` math.Exp(math.Log(p.Xm) - p.Alpha * rnd.ExpFloat64())` (I think that's the right sign)

File: stat/distuv/pareto.go [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+func (p Pareto) StdDev() float64 {
+	return math.Sqrt(p.Variance())
+}
+
+// Survival returns the survival function (complementary CDF) at x.
+func (p Pareto) Survival(x float64) float64 {
+	if x < p.Xm {
+		return 1
+	}
+	return math.Pow(p.Xm/x, p.Alpha)
```
Comment (btracey) [inline stat/distuv/pareto.go]: I would also do this in log space and convert it back. `math.Exp(p.Alpha*( math.Log(p.Xm) - math.Log(x)))`

File: stat/distuv/pareto.go [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+// CDF computes the value of the cumulative density function at x.
+func (p Pareto) CDF(x float64) float64 {
+	if x < p.Xm {
+		return 0
+	}
+	return 1 - math.Pow(p.Xm/x, p.Alpha)
+}
+
```
Comment (btracey) [inline stat/distuv/pareto.go]: Entropy?

File: stat/distuv/pareto_test.go [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+			t.Errorf("Pdf mismatch, x = %v, xm = %v, alpha = %v. Got %v, want %v", test.x, test.xm, test.alpha, pdf, test.want)
+		}
+	}
+}
+
+func TestParetoCDF(t *testing.T) {
+	for _, test := range []struct {
+		x, xm, alpha, want float64
+	}{
+		{0, 1, 1, 0},
```
Comment (btracey) [inline stat/distuv/pareto_test.go]: Try some non-integer values in here too please.

Comment (kczimm) [reply stat/distuv/pareto_test.go]: done.

File: stat/distuv/pareto.go (hunk 12 above)
Comment (kczimm) [reply stat/distuv/pareto.go]: done.

File: stat/distuv/pareto.go [hunk 17]
```diff
@@ -34,7 +34,12 @@ func (p Pareto) CDF(x float64) float64 {
 	if x < p.Xm {
 		return 0
 	}
-	return 1 - math.Pow(p.Xm/x, p.Alpha)
+	return 1 - p.Survival(x)
+}
+
+// Entropy returns the differential entropy of the distribution.
+func (p Pareto) Entropy() float64 {
+	return math.Log((p.Xm / p.Alpha) * math.Exp(1+1/p.Alpha))
```
Comment (btracey) [inline stat/distuv/pareto.go]: `math.Log(p.Xm) - math.Log(p.Alpha) + (1 + 1/p.Alpha)`

**Comment by kczimm** [comment]
Fixed entropy and squashed.

**Pull request merged**